0101-nix formatting.md

```diff
@@ -0,0 +1,672 @@
+  produces a diff in two elements:
+
+  ```diff
+  - { lib
+  - , mkDerivation
+  + { mkDerivation
+    , fetchurl
+    , ...
+    } @ attrs:
+    stdenv.mkDerivation # ...
+  ```
+
+- Documenting the first argument creates an inconsistency
+  between the way arguments start:
+
+  ```nix
+  {
+    # Lorem Ipsum
+    lib
+  , mkDerivation
+  , fetchurl
+  , ...
+  } @ attrs:
+  stdenv.mkDerivation # ...
+  ```
+
+- This is not consistent with _Let-In_, and _AttrSet_,
+  where the separator goes after the element
+  instead of at the beginning.
+- It ruins "folding by indentation" modes
+  on Vim, Neovim, VSCode, and other major code editors,
+  because the data-structure has the same indentation
+  as the opening brace.
+
+❌ Bad:
+
+```nix
+{ mkDerivation, lib, fetchurl, ... }@attrs: stdenv.mkDerivation # ...
+```
+
+- One-liners are unreadable and produce bad diffs.
+
+❌ Bad:
+
+```nix
+{ mkDerivation, lib, fetchurl, extra-cmake-modules, kdoctools, wrapGAppsHook
+, karchive, kconfig, kcrash, kguiaddons, kinit, kparts, kwind, ... }@attrs:
+stdenv.mkDerivation # ...
+```
+
+- It's hard to tell this destructuring has an ellipsis (`...`) at a first glance,
+  because it's mixed with the other arguments.
+- Moving elements becomes harder
+  than a simple whole-line movement.
+  (Moving a whole line is normally a keyboard-shortcut
+  or command in major code editors).
+- Excessively compact:
+  adding, removing, or editing an argument
+  produces a diff in more than one argument.
+- `}@attrs` is not intuitive
+  with the rules of written english,
+  where you add whitespace
+  after the end of the previous phrase
+  (`phrase. Other phrase`).
+
+### Inherit
+
+✅ Good:
+```nix
+inherit foo bar;
+inherit (attrs) foo bar;
+inherit
+  foo
+  bar
+  ;
+inherit (attrs)
+  foo
+  bar
+  ;
+```
+
+- Inherit expressions can have their symbols either all on the same line as the `inherit`, or one line per element each.
+- The `(attrs)` is always on the same line as the `inherit` statement. There are no spaces on the inside of its parentheses.
+
+❌ Bad:
+```nix
+TODO
+inherit (attrs) foo
+  bar
+  ;
+```
+
+❌ Bad:
+```nix
+inherit
+  (attrs)
+  foo
+  bar
+  ;
+```
+
+✅ Good:
+
+```nix
+{
+  inherit
+    (mkFruit {
+      a = "apple";
+      b = "banana";
+      c = "cherry";
+      d = "date";
+      e = "elderberry";
+    })
+    feijoa
+    grape
+    honeydew
+    ;
+}
+```
+
+- Technically this is inconsistent because the attribute set is sometimes on the same line as the `inherit` and other times not. However, this is a really niche use case and generally you won't need to write such code anyways so we are willing to sacrifice consistency for a better default.
+
+### If-Then-Else
+
+✅ Good:
+
+```nix
+if predicate then
+  foo
+else
+  bar
+```
+
+- "Pulling up" the `then` keyword is consistent with other similar syntax constructs
+- The bodies are always indented
+
+❌ Bad (old):
+
+```nix
+if predicate
+then foo
+else bar
+```
+
+- TODO explain why bad
```

Comment:
I think a good reasoning would include multi-line expressions inside the then and else blocks:

```nix
if something <= 4.0
then
   {
     foo = 10;
   }
else if something <= 6.0
then
   [
     10
     20
   ]
else if something <= 7.0
then
   let
     …
   in
     …
```

Is overly long and has no advantage.

```nix
if something <= 4.0
then {
     foo = 10;
   }
else if something <= 6.0
then [
     10
     20
   ]
else if something <= 7.0
then let
     …
   in
     …
```

Is just as long as having `then` in line, but influences all the other formatting rules and it’s harder to tell at a glance where the blocks start and what they are.

`then` is really just a keyword for the parser, for the human reading it’s just noise. `if` and `else` are the important parts of conditionals.